feat: upgrade to new fastapi, update models to handle both pydantic v…

timothyjlaurent opened the pull request:
…1 an v2, add matrix conditions for pydantic_v1 and pydantic_v2

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kserve/kserve/blob/master/docs/DEVELOPER_GUIDE.md
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This work builds on this PR https://github.com/kserve/kserve/pull/3273 that adds pydantic 2 support to the kserve package.

The following is copied from that PR:
* Follows Optional fields guidance for pydantic v2 [ref](https://docs.pydantic.dev/latest/migration/#required-optional-and-nullable-fields)
*Overwrites protected_namespace in pydantic v2 to get rid of warning logs about namespace conflicts in pydantic [ref](https://docs.pydantic.dev/latest/api/config/#protected-namespaces). An example is shown as follows
```
E   UserWarning: Field "model_name" has conflict with protected namespace "model_".
E
E   You may be able to resolve this warning by setting `model_config['protected_namespaces'] = ()`.
```

In addition to handling those incompatibilities it also allows for both pydantic 1.x and 2.x versions.

To accomplish this:
- FastAPI is upgraded to the latest version. This version supports both pydantic 1.x and 2.x. As such, projects that add kserve as a dependency will also be able to use both pydantic versions.
- inspired by how https://github.com/docarray/docarray added support for v1 and v2 pydantic an `is_pydantic_2` flag is used to change the underlying pydantic implementations
- each pydantic model that has incompatible syntax, ie nested `Config` class in retains the class if it is pydantic 1 and uses the `model_config` syntax if it is pydantic 2. Keys are changed where needed, eg `schema_extra` -> `json_schema_extra`.
- `protected_namespace=()` is set on the `InferenceResponse` model
- to test, again inspired by docarray's CI, a new matrix component `pydantic-version` is added to CI. As the default version will be pydantic 2, in the case that the version is `pydantic_1` then the make target `install_pydantic_1` is executed which downgrades pydantic to v1.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #3373 

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Added matrix to test both pydantic 1 and pydantic 2

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Upgrades FastAPI version and now supports Pydantic v1.x and v2.x. This allows dependent projects to upgrade their Pydantic version as needed.
```

terrytangyuan commented on .github/workflows/python-test.yml:
```diff
@@ -35,6 +35,7 @@ jobs:
     strategy:
       matrix:
         python-version: ["3.8", "3.9", "3.10", "3.11"]
+        pydantic-version: ["pydantic_1", "pydantic_2"]
```
It's great to see this coverage but I feel like this could significantly increase the build time since we have to test the combo of Python and Pydantic versions.

timothyjlaurent replied:
Ah good point - especially since we don't need to test pydantic-v1 on all of the other python packages.

I can work on making another set of steps for testing the pydantic v1 functionality.

timothyjlaurent replied:
I've added steps for testing this in the main build job and removed the extra matrix arguments

sivanantha321 commented on python/kserve/pyproject.toml:
```diff
@@ -35,7 +35,7 @@ exclude = ["kserve/storage/test"]
-fastapi = "^0.95.0"
+fastapi = "^0.109.0"
 cloudevents = "^1.6.2"
 six = "^1.16.0"
```
I think it is better to explicitly add pydantic dependency here.

sivanantha321 replied:
Model servers depends on kserve sdk. So to propagate the version bump, you have to update the lockfiles of serving runtimes. Running `hack/python-releash.sh` would help.

sivanantha321 replied:
One more thing, I recently found out that fastapi 0.109.0 breaks ray serve. https://github.com/ray-project/ray/pull/42740. So let's keep it as it is for now.

timothyjlaurent replied:
Well to upgrade pydantic to v2 we need to upgrade to at least 0.100.0.

https://github.com/tiangolo/fastapi/releases/tag/0.100.0 -- it's probably best to increment it maximally so I'll set it to 0.108.0 as was done in the linked PR.

timothyjlaurent replied:
@sivanantha321 I updated the lockfiles with the serving runtimes, added the pydantic 2 syntax to the new models, and added an explicit pydantic dependency.

sivanantha321 commented:
@timothyjlaurent Can you help rebase the PR and sign off the commits ?

timothyjlaurent commented:
@sivanantha321 I've rebased and signed all the commits, can you take another look?

timothyjlaurent committed:
feat: add support for pydantic v2, tests for pydantic v1, and update dependencies

Signed-off-by: Timothy Laurent <tlaurent@genedx.com>

sivanantha321 commented:
@yuzisun Can you help trigger the tests ?

timothyjlaurent committed:
feat: regenerate lock files

Signed-off-by: Timothy Laurent <tlaurent@genedx.com>

timothyjlaurent commented:
We had a small bug in the last test and then I somehow borked my rebase -- I've cleaned up the commits on a fresh branch and think it should pass now.

sivanantha321 commented:
@timothyjlaurent I don't see a updated version of pydantic in poetry.lock file ? I think we can upgrade the pydantic version to >2.0.0 for model servers.

timothyjlaurent commented:
@sivanantha321 Ah this is because we hack/python-release runs with --no-update and the sdk is compatible with pydantic v1 or v2.

I can either run it to update all packages, or I could add pydantic as an explicit dep to each model server. What do you think?

sivanantha321 commented:
> @sivanantha321 Ah this is because we hack/python-release runs with --no-update and the sdk is compatible with pydantic v1 or v2.
> 
> 
> 
> I can either run it to update all packages, or I could add pydantic as an explicit dep to each model server. What do you think?

You can update the packages as long as it does not breaks anything

timothyjlaurent committed:
feat: regenerate lock files with updates

Signed-off-by: Timothy Laurent <tlaurent@genedx.com>

timothyjlaurent commented:
@sivanantha321 I've updated the deps.

yuzisun commented:
@timothyjlaurent Thanks for upgrading pydantic! Can you help take a look a look at the testing failures?
https://github.com/kserve/kserve/actions/runs/7993801929/job/21947239919?pr=3374

gujingit commented on python/kserve/kserve/protocol/rest/v2_datamodels.py:
```diff
@@ -111,303 +132,376 @@ class ModelMetadataResponse(BaseModel):
+
     text_output: str
     model_name: str
     model_version: Optional[str]
```
```
model_version: Optional[str] = None
details: Optional[Details] = None
```

timothyjlaurent replied:
Thanks, I've updated all of the optionals

gujingit commented on python/kserve/pyproject.toml:
```diff
@@ -35,7 +35,7 @@ exclude = ["kserve/storage/test"]
 [tool.poetry.dependencies]
 python = ">=3.8,<3.12"
 uvicorn = { version = "^0.19.0", extras = ["standard"] }
```
uvicorn = { version = "^0.21.1", extras = ["standard"] }

gujingit replied:
@timothyjlaurent Can you help update uvicorn version to "^0.21.1" ？So that PR #3480 can be closed.

timothyjlaurent replied:
Sure thing, added it and updated all dependent packages

timothyjlaurent committed:
fix: update dependencies, fix broken tests

timothyjlaurent committed:
feat: add special case for testing model ready, "True" for pydantic v1

timothyjlaurent commented:
> @timothyjlaurent Thanks for upgrading pydantic! Can you help take a look a look at the testing failures? https://github.com/kserve/kserve/actions/runs/7993801929/job/21947239919?pr=3374

I took the time to figure out how to properly run the test suite locally -- needs to be done from the ./python directory.

I've updated the tests one notable workaround is that the RayServe test asserted `b'{"name":"TestModel","ready":"True"}'` <- notice the stringified bool. In pydantic V2 we correctly get an actual json bool for the ready param. I added a special case to the test and left comments explaining this behavior.

sivanantha321 commented on python/kserve/test/test_server.py:
```diff
@@ -381,7 +382,14 @@ def test_list_handler(self, http_server_client):
-        assert resp.content == b'{"name":"TestModel","ready":"True"}'
+        # for some reason the RayServer responds with the stringified python bool
+        # when run on pydantic < 2 and the bool when run on pydantic >= 2
+        # eg {"name":"TestModel","ready":"True"} vs {"name":"TestModel","ready":true}
```
I think it is a different bug altogether. Before moving to fastapi, it seems the [health handler returned bool values](https://github.com/kserve/kserve/blob/21ce62bf4522e63ff6a60435250c5cbd55b12b43/python/kserve/kserve/handlers/health.py#L25-L47), 
but after the migration it started [returning string values because how the type is defined](https://github.com/kserve/kserve/blob/e6eede2ea0e94cbe75e90785877cba33f9014216/python/kserve/kserve/protocol/rest/v1_endpoints.py#L39) ?
@yuzisun

yuzisun replied:
The code seems to be returning bool, as @timothyjlaurent mentioned the issue is for the rayserve case?

timothyjlaurent replied:
This is for the RayServe case.

In the previous assertion with pydantic v1 - we were getting `{"name":"TestModel","ready":"True"}` the string value, and testing for that. Seems like it was a bug that got "blessed" by enshrinining it in a testcase.
With the pyanticv2 we're (correctly) getting a bool value for the `ready` key (not a stringified python string) - This, I believe is the correct behavior we should expect.

yuzisun commented:
@timothyjlaurent seems like it keeps failing after updating the poetry lock file
https://github.com/kserve/kserve/actions/runs/8087460553/job/22214167194?pr=3374

yuzisun committed:
Merge branch 'master' into 3373-support-pydantic-2

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun committed:
Update poetry lock

Signed-off-by: Dan Sun <dsun20@bloomberg.net>

yuzisun commented:
Tests passed, Thanks @timothyjlaurent for the great work!

/lgtm
/approve
